### PR TITLE
Fix db upload issue

### PR DESCRIPTION
**Pull request by fsi206914**
When too many projects are being uploaded simultaneously, we find an issue that the mysql server can not work properly such that memory starvation happens.  We enforce az committing to db when uploading every single chunk, in order to resolve this problem.

Already verified in our testing cluster.

### CODE REVIEW

**Comment by HappyRay** [comment]
In 
public void uploadProjectFile(Project project, int version, String filetype,
      String filename, File localFile, String uploader)
      throws ProjectManagerException {
    logger.info("Uploading to " + project.getName() + " version:" + version
        + " file:" + filename);
    Connection connection = getConnection();

    try {
      uploadProjectFile(connection, project, version, filetype, filename,
          localFile, uploader);
      connection.commit();
      logger.info("Commiting upload " + localFile.getName());
    } catch (SQLException e) {
      logger.error(e);
      throw new ProjectManagerException("Error getting DB connection.", e);
    } finally {
      DbUtils.closeQuietly(connection);
    }
  
Does it make sense to do
connection.rollback();
in the catch block?
See 
http://stackoverflow.com/questions/15761791/transaction-rollback-on-sqlexception-using-new-try-with-resources-block
This may be important when we implement connection pooling.

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java [hunk 1]
```diff
@@ -382,6 +382,25 @@ private void uploadProjectFile(Connection connection, Project project,
     }
 
     logger.info("Md5 hash created");
+
+    /**
+     * Insert a version row to table project_versions.
+     * If something goes wrong during uploading project files to DB, the latest version number
+     * in project_versions refers to the destructive project_file. Since AZ always let the number+1
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Do you mean "incomplete project files?"

Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Do you mean " this is desirable to ensure that the next version number will be unique"?

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java [hunk 2]
```diff
@@ -416,16 +438,20 @@ private void uploadProjectFile(Connection connection, Project project,
       IOUtils.closeQuietly(bufferedStream);
     }
 
-    final String INSERT_PROJECT_VERSION =
-        "INSERT INTO project_versions (project_id, version, upload_time, uploader, file_type, file_name, md5, num_chunks) values (?,?,?,?,?,?,?,?)";
+    /**
+     * Since we initialized num_chunks to 0 before uploading files, we set it
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Not sure I follow the comment.

**Comment by HappyRay** [comment]
Consider explaining how this change would solve the problem and how it handles error cases.

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java [hunk 3]
```diff
@@ -416,16 +438,20 @@ private void uploadProjectFile(Connection connection, Project project,
...
-          version, updateTime, uploader, filetype, filename, md5, chunk);
+      runner.update(connection, UPDATE_PROJECT_NUM_CHUNKS, chunk, project.getId(), version);
+      connection.commit();
     } catch (SQLException e) {
-      logger.error(e);
-      throw new ProjectManagerException("Error updating project version "
-          + project.getName(), e);
+      throw new ProjectManagerException(
+          "Error updating project file chunk_num " + project.getId() + ":"
+              + chunk, e);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Nice improvement to the logging. Thanks.

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java [hunk 4]
```diff
@@ -382,6 +382,25 @@ private void uploadProjectFile(Connection connection, Project project,
     }
 
     logger.info("Md5 hash created");
+
+    /**
+     * Insert a version row to table project_versions.
+     * If something goes wrong during uploading project files to DB, the latest version number
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Do you mean " something goes wrong after this entry is inserted"?

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java (hunk 1 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Explain why this update goes first before inserting the data chunks?

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java [hunk 5]
```diff
@@ -402,9 +421,12 @@ private void uploadProjectFile(Connection connection, Project project,
           logger.info("Running update for " + filename + " chunk " + chunk);
           runner.update(connection, INSERT_PROJECT_FILES, project.getId(),
               version, chunk, size, buf);
+
+          // We enforce az committing to db when uploading every a single chunk.
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Explain the reason. 
How about " to reduce the transaction duration and conserve sql server resources"?

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java (hunk 4 above)
Comment (fsi206914) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Fixed.

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java (hunk 1 above)
Comment (fsi206914) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Fixed.

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java (hunk 5 above)
Comment (fsi206914) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Fixed.

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java (hunk 2 above)
Comment (fsi206914) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Fixed.

**Pull request merged**